src/hypothesis/internal/compat.py

```diff
@@ -502,10 +502,16 @@ def implements_iterator(it):
 
 if PY2:
     def floor(x):
-        return int(math.floor(x))
+        # `int(math.floor(x))` fails for negative long integers
```

Comment:
```
@given(fractions())
def test_floor_ceil(x):
    # fails with `test_floor_ceil(x=Fraction(9007199254740993, 1))`
    # ==> `assert 9007199254740993L == 9007199254740992L`
    assert ceil(x) == int(math.ceil(x))
    # similar for `assert floor(x) == int(math.floor(x))`
```

I suspect `math.ceil` is casting via float at some point, but the why doesn't really matter - this is a real bug that causes my fractions tests to fail 😄

The slightly longer implementation overflows to Python2 long integers, which keep working until you run out of memory - without any drift at all.